### PR TITLE
Adding subscription config into subscription job.

### PR DESCRIPTION
**Pull request by n3wscott**
fixes https://github.com/knative/eventing/issues/1424

### CODE REVIEW

File: pkg/apis/pubsub/v1alpha1/pull_subscription_validation.go [hunk 1]
```diff
@@ -49,6 +55,14 @@ func (current *PullSubscriptionSpec) Validate(ctx context.Context) *apis.FieldEr
 			errs = errs.Also(err.ViaField("transformer"))
 		}
 	}
+
+	if current.RetentionDuration != nil {
+		// If set, RetentionDuration Cannot be longer than 7 days or shorter than 10 minutes.
+		if *current.RetentionDuration < minRetentionDuration || *current.RetentionDuration > maxRetentionDuration {
+			errs = errs.Also(apis.ErrInvalidValue(current.RetentionDuration, "retentionDuration"))
```
Comment (mattmoor) [inline pkg/apis/pubsub/v1alpha1/pull_subscription_validation.go]: We have an OOB error function you should use for bounded values.

File: pkg/pubsub/operations/subscription.go [hunk 2]
```diff
@@ -56,7 +63,23 @@ func NewSubscriptionOps(arg SubArgs) *batchv1.Job {
...
+		env = append(env, []corev1.EnvVar{{
+			Name:  "PUBSUB_SUBSCRIPTION_CONFIG_ACK_DEAD",
+			Value: arg.AckDeadline.String(),
+		}, {
+			Name:  "PUBSUB_SUBSCRIPTION_CONFIG_RET_ACKED",
+			Value: strconv.FormatBool(arg.RetainAckedMessages),
+		}, {
+			Name:  "PUBSUB_SUBSCRIPTION_CONFIG_RET_DUR",
+			Value: arg.RetentionDuration.String(),
+		}}...)
```
Comment (mattmoor) [inline pkg/pubsub/operations/subscription.go]: It'd be nice if `envconfig` had a method for creating this from the structured type :-/

File: pkg/apis/pubsub/v1alpha1/pull_subscription_defaults.go [hunk 3]
```diff
@@ -43,9 +47,20 @@ func defaultSecretSelector() *corev1.SecretKeySelector {
 }
 
 func (ss *PullSubscriptionSpec) SetDefaults(ctx context.Context) {
+	if ss.AckDeadline == nil {
+		ackDeadline := defaultAckDeadline
+		ss.AckDeadline = &ackDeadline
```
Comment (mattmoor) [inline pkg/apis/pubsub/v1alpha1/pull_subscription_defaults.go]: Add `ptr.Duration`?

File: pkg/pubsub/operations/subscription.go [hunk 4]
```diff
@@ -128,26 +172,49 @@ func (s *SubscriptionOps) Run(ctx context.Context) error {
 		logger.Info("Previously created.")
 
 	case ActionCreate:
+		// Load the topic.
+		topic, err := s.getTopic(ctx)
+		if err != nil {
+			logger.Fatal("Failed to get topic.", zap.Error(err))
```
Comment (mattmoor) [inline pkg/pubsub/operations/subscription.go]: Seems really bad to `logger.Fatal` in a library.

File: pkg/pubsub/operations/subscription.go [hunk 5]
```diff
@@ -128,26 +172,49 @@ func (s *SubscriptionOps) Run(ctx context.Context) error {
...
+			// Compare the current config to the expected config. Update if different.
+			if diff := cmp.Diff(subConfig, currentConfig, ignoreSubConfig); diff != "" {
+				_, err := sub.Update(ctx, pubsub.SubscriptionConfig{
+					AckDeadline:         s.AckDeadline,
+					RetainAckedMessages: s.RetainAckedMessages,
+					RetentionDuration:   s.RetentionDuration,
+					Labels:              currentConfig.Labels,
+				})
+				if err != nil {
+					logger.Fatal("Failed to update subscription config.", zap.Error(err))
```
Comment (mattmoor) [inline pkg/pubsub/operations/subscription.go]: same.

File: pkg/pubsub/operations/subscription.go (hunk 4 above)
Comment (n3wscott) [reply pkg/pubsub/operations/subscription.go]: This kills the job and marks it failed. working as intended.

File: pkg/pubsub/operations/subscription.go (hunk 5 above)
Comment (n3wscott) [reply pkg/pubsub/operations/subscription.go]: This kills the job and marks it failed. working as intended.

File: pkg/apis/pubsub/v1alpha1/pull_subscription_defaults.go (hunk 3 above)
Comment (n3wscott) [reply pkg/apis/pubsub/v1alpha1/pull_subscription_defaults.go]: https://github.com/knative/pkg/pull/519/

File: pkg/pubsub/operations/subscription.go (hunk 4 above)
Comment (n3wscott) [reply pkg/pubsub/operations/subscription.go]: ok I will fix.

Comment (n3wscott) [reply pkg/pubsub/operations/subscription.go]: Fixed.

File: pkg/apis/pubsub/v1alpha1/pull_subscription_validation.go (hunk 1 above)
Comment (n3wscott) [reply pkg/apis/pubsub/v1alpha1/pull_subscription_validation.go]: Found it, used it.

File: pkg/pubsub/operations/subscription.go (hunk 2 above)
Comment (n3wscott) [reply pkg/pubsub/operations/subscription.go]: I made an issue to track this idea.

File: pkg/reconciler/pullsubscription/pullsubscription.go [hunk 6]
```diff
@@ -210,7 +210,7 @@ func (c *Reconciler) reconcile(ctx context.Context, source *v1alpha1.PullSubscri
 
 	source.Status.SubscriptionID = resources.GenerateSubscriptionName(source)
 
-	state, err := c.EnsureSubscriptionCreated(ctx, source, source.Spec.Project, source.Spec.Topic, source.Status.SubscriptionID)
+	state, err := c.EnsureSubscriptionCreated(ctx, source, source.Spec.Project, source.Spec.Topic, source.Status.SubscriptionID, *source.Spec.AckDeadline, source.Spec.RetainAckedMessages, *source.Spec.RetentionDuration)
```
Comment (Harwayne) [inline pkg/reconciler/pullsubscription/pullsubscription.go]: This is getting long enough that it should probably be a struct.

Comment (n3wscott) [reply pkg/reconciler/pullsubscription/pullsubscription.go]: https://github.com/GoogleCloudPlatform/cloud-run-events/issues/101

**Comment by n3wscott** [comment]
track the optional comment with https://github.com/GoogleCloudPlatform/cloud-run-events/issues/101

I think that is a good cleanup but does not change the functionality. 

/hold cancel

**Pull request merged**